framework/src/media/audio/audio_manager.c

```diff
@@ -181,54 +253,124 @@ audio_manager_result_t find_output_audio_card(void)
 	return AUDIO_MANAGER_SUCCESS;
 }
 
-audio_manager_result_t set_audio_volume(uint16_t volume)
+static audio_manager_result_t set_audio_volume(audio_device_type_t type, uint8_t volume)
 {
 	int ret;
-	struct pcm *pcm;
-
-	if ((ret = get_active_audio_device_pcm(&pcm, OUTPUT)) != AUDIO_MANAGER_SUCCESS) {
-		return ret;
-	}
-
+	int fd;
 	struct audio_caps_desc_s caps_desc;
 	FAR struct audio_caps_s *caps = &(caps_desc.caps);
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_GETCAPS, (unsigned long)caps);
-	if (ret < 0) {
-		auddbg("Fail to ioctl AUDIOIOC_CONFIGURE, ret = %d\n", ret);
+	audio_config_t *audio_config = (type == OUTPUT ? & (g_audio_out_cards[g_actual_audio_out_card_id].config) : & (g_audio_in_cards[g_actual_audio_in_card_id].config));
+	char *path = (type == OUTPUT ? g_audio_out_cards[g_actual_audio_out_card_id].card_path : g_audio_in_cards[g_actual_audio_in_card_id].card_path);
+
+	if ((fd = (type == OUTPUT ? open(path, O_WRONLY) : open(path, O_RDONLY))) < 0) {
+		auddbg("Fail to open driver, path = %s\n", path);
 		return AUDIO_MANAGER_DEVICE_FAIL;
 	}
 
-	caps->ac_controls.hw[0] = volume;
+	if (volume > AUDIO_DEVICE_MAX_VOLUME) {
+		volume = AUDIO_DEVICE_MAX_VOLUME;
+	}
 
+	caps->ac_len = sizeof(struct audio_caps_s);
 	caps->ac_type = AUDIO_TYPE_FEATURE;
-	caps->ac_format.hw = AUDIO_FU_VOLUME;
 
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_CONFIGURE, (unsigned long)&caps_desc);
+	if (type == INPUT) {
```

Comment:
Here we already have a if-else, then let's do all branch code here.

Reply:
Refactored.